genesis/man/man1/convert.1
.\" $Id: convert.1,v 1.4 2005/07/01 20:20:52 svitak Exp $
.TH CONVERT 1L "3 October 1994"
.SH NAME
convert \- convert GENESIS 1.x scripts to GENESIS 2.x syntax and command names
.SH SYNOPSIS
.B convert
[
.I oldscriptfile
[
.I newscriptfile
]
]
[
.B \-noX1compat
]
[
.B \-constants
.I scriptfile ...
]
.SH DESCRIPTION
.LP
.I Convert
converts a GENESIS 1.x script file
to GENESIS 2.x syntax replacing old command names and arguments
with their GENESIS 2.x counterparts.
Without arguments
.I convert
will read standard input and write to standard output.
With arguments
.I convert
reads from and writes to the given file names.
If the
.I newscriptfile
argument is not given,
.I convert
will output to a file with the same name as the input
file less any directory path.
If a directory path is not given for the input file
output is written to standard output.
.LP
To convert entire script sets
UNIX command shell loops may be used.
Assuming that an old set of scripts
lives in the directory ../oldscripts,
then the following will convert the entire script set.
.LP
For csh:
.nf
.sp
foreach file (../oldscripts/*.g)
    convert $file
end
.LP
For sh:
.nf
.sp
for file in ../oldscripts/*.g
do
    convert $file
done
.SH OPTIONS
.LP
.B \-constants
.LP
To correctly translate a script
.I Convert
needs to know the names of globally defined variables
in a script set.
The
.B \-constants
option is used to identify the scripts
in which these global script variables are defined.
More than one script may be specified.
.LP
.B \-noX1compat
.LP
The automatic translation of create commands for Xodus objects
to objects in the X1compat library may be suppressed using the
.B \-noX1compat
option.
In this case, converted scripts will use the Xodus 2 objects
directly.  This will work most of the time for the standard
Xodus 1 interface objects, but will likely fail for the
draw widgets and pixes.  xgraph widgets may also fail when
not using X1compat but have a better chance of working than
xdraw et al.
.SH "SEE ALSO"
The GENESIS 1.x to GENESIS 2.x conversion section
of the GENESIS Reference Manual.
.SH BUGS
.LP
.I Convert
is not a complete conversion solution.
See the Doc/convert.doc help file for a
complete discussion of conversion issues
including a list of problem areas.
.LP
.I Convert
will delete the contents of the input file
if the output filename is the same as the input file name.
This can be done implicitly if the input filename path
refers to a file in the current directory using a relative path
(e.g. convert ./file.g).
.LP
.I Convert
makes use of the X1compat script set
for compatibility with GENESIS 1.x graphical interface elements.
The X1compat script directory should be added to the SIMPATH
and the X1compat script called from the .simrc
or the main simulation script.
.LP
.I Convert
translates network models using GENESIS 1.x connections
to commands in the connection compatibility library.
This library is not included by default in the GENESIS 2.x
executable.
See the GENESIS Reference Manual for details on compiling
GENESIS 2.x with the connection compatibility library
and for tips on rewriting network simulations using the
new network modeling objects and commands.


genesis/ChangeLog
                             G E N E S I S
                Summary of changes and revision history
                ***************************************
                Last edited: $Date: Tue Nov 11 14:05:40 MST 2014 $

Each set of changes listed below is prefaced with the version number for
which the changes apply.  For development releases, two version numbers
are listed on separate lines, the first being the development release
version and the second being the proposed public release version.  The
proposed release is subject to discussion and change.


------------------------------------------
FINAL RELEASE 2.4
2014-November
------------------------------------------

- A new autoconf-based build system has been incorporated, so that a
simple "./configure" command will create the appropriate Makefile.

- Three new network connection commands, 'volumeconnect3', 'volumeweight3',
and 'volumedelay3', have been added and documented. These allow
distance-dependent connection probabilities and/or weights and the use
of periodic boundary conditions for network connectivity.

- Added HTML indexing to the 'genesis' directory and subdirectories.
README files have been replaced or supplemented with '.txt' or '.html' files.

- Several source code bug-fixes and changes to src/Makefile.dist were
made in order to solve compilation problems on newer OS versions.

------------------------------------------
BETA RELEASE 2.4 beta
2014-July
------------------------------------------

- addition of 'stdp_rules' object, documentation, and Scripts for implementing
spike timing dependent plasticity in networks of hsolved cells

- the chemesis library for modeling second messengers and calcium dynamics
has been added to GENESIS

- new tutorial Scripts were added for chemesis, gpython-tools,
purkinje, stdp_rules, VAnet2

------------------------------------------
FINAL RELEASE 2.3
2006-March
------------------------------------------
-Last-minute pgenesis doc updates.

-Makefile.dist changes to development targets srcdist and bindist and some
comment clarifications.

-Added Izhikevich simple spiking neuron model objects and demo scripts. Plain
text documentation is included under Doc only.

-Added float_format field for asc_file and res_ascfile objects.

-Update of the new purkinje cell tutorial to allow graphing of multiple data
streams.

-Documentation has been revised for BoG.txt, Changes.txt, asc_file.txt, and 
Objects.txt, tabchannel.txt, README.cygwin.

-Some files were moved and Makefile.dist updated to facilitate creating src and
binary distributions with minimal impact on the existing directory hierarchy.


------------------------------------------
STABLE RELEASE CANDIDATE 2.3
2006-January
------------------------------------------
- Completely updated documentation! Including pgenesis and many of the
  tutorials.

- Fixed some display issues in the squid tutorial and eliminated the restriction
  on the number of steps which can be performed.

- Fixed crash of Xodus when resizing windows.

- Made Xodus dialogs behave correctly in the face of newer window managers
  which may attempt to enforce "focus stealing prevention".

- Script parser should now operate correctly on scripts with Mac- and Windows/
  DOS- like line endings.

- A new purkinje tutorial with a cell plugin framework, on-the-fly generation
  of tabchannels, and the ability to select compartment names for ascii ouput
  and current injection.

- Added channels and made some small changes to accomodate leechlib.

- Fixed small bug in Xodus display.

- Increased sizes of buffers for compartment names to accomodate cvapp
  conversion from neurolucida files.

- Updated kinetikit example files to sync with kkit11.

- Additional makefile targets for binary distributions.
  Intended for developers.


------------------------------------------
DEVELOPMENT RELEASE 2.3 (aka 2.3 beta)
All changes intended for 2.3 release
2005-August
------------------------------------------

- Incorporated patches for variable argument functions.
  The old-style varargs.h usage is deprecated and has been replaced by
  stdarg.h.

- Incorporated MacOSX patches submitted by Alfonso Delgado-Reyes,
  Cengiz Gunay, and Werner Van Geit. Since OSX (and other OS's) may use
  case-insensitive file systems, some directory name conflicts
  were resolved under the Xodus tree.

- Incorporated Cygwin patches submitted by Sergio Daicz.
  Cygwin installations had the same problems mentioned above with Xodus
  directory names requiring the same solution. Other fixes included
  changing the size of hash table to avoid overflow. This was fixed by
  changing the hash table size to a prime number (10007), resulting in a
  more evenly distributed hash.

- Incorporated bug fixes for the kinetics library.

- Removed references to PARSER in Makefiles since it is no longer used.
  Since it's use was not well documented, it was tempting to define it
  with the same path as YACC. This would cause problems on most compilers
  since PARSER was used as an argument to '-D'.

- Removed [references to] extra fortran and C++ source code files.
  The fortran and c++ code libraries are not used by GENESIS.

- Modified Makefiles to pass appropriate variables to SPRNG.
  The SPRNG random number generation libraries should now compile on
  all known architectures. Variables are now passed from the top-level
  Makefile.


------------------------------------------
STABLE RELEASE 2.2.1
2005-June
------------------------------------------

* 2.2.1 development code relocated to sourceforge repository

- Changed AUTHORS permissions to 644 from 755

- Removed empty file VERSION_2.2.1

- Removed empty directory src/sprng/DOCS

- Added COPYRIGHT

- Added GLPLicense, GNU Lesser Public License

- Removed file src/diskio/interface/netcdf/netcdf-3.4/src/macros.make
  since it is generated from macros.make.in

- Removed Doc/Manual.tar.Z. Now accessible from the sourceforge

genesis/src/convert/convert.c
static char rcsid[] = "$Id: convert.c,v 1.3 2005/07/01 10:03:01 svitak Exp $";

/*
** $Log: convert.c,v $
** Revision 1.3  2005/07/01 10:03:01  svitak
** Misc fixes to address compiler warnings, esp. providing explicit types
** for all functions and cleaning up unused variables.
**
** Revision 1.2  2005/06/27 19:00:33  svitak
** Added explicit types to untyped functions and fixed return values as
** appropriate. Initialized values explicitly when it was unclear if
** they were being used uninitialized (may not apply to all files being
** checked in in this batch).
**
** Revision 1.1.1.1  2005/06/14 04:38:28  svitak
** Import from snapshot of CalTech CVS tree of June 8, 2005
**
** Revision 1.5  2001/04/18 22:39:36  mhucka
** Miscellaneous portability fixes, mainly for SGI IRIX.
**
** Revision 1.4  1999/12/29 10:24:59  mhucka
** Fixed compilation warning involving signal return datatypes.
**
** Revision 1.3  1997/07/18 03:13:37  dhb
** Fix for getopt problem; getopt(), optopt and optind are now
** G_getopt(), G_optopt and G_optind.
**
** Revision 1.2  1995/02/01 23:01:03  dhb
** Added support for toggling use of X1compat library using
** -noX1compat switch.
**
 * Revision 1.1  1995/01/13  01:09:48  dhb
 * Initial revision
 *
*/

#include <stdio.h>
#include <string.h>
#include "system_deps.h"
#include "header.h"

extern int yydebug;
extern int InputFs();
extern int yyparse();

static char* CurrentFile = "<stdin>";

void yyerror(str)
char*	str;
{
	fprintf(stderr, "%s: error: %s\n", CurrentFile, str);
	fprintf(stderr, "conversion failed\n");
	exit(1);
}

void yyaccept()
{
	exit(0);
}

SIGTYPE
sig_msg_restore_context()
{
	fprintf(stderr, "%s: conversion failed\n", CurrentFile);
	exit(1);
}

int main(argc, argv)
int	argc;
char*	argv[];
{	/* convert --- Convert a script from GENESIS 1.4.1 to 2.0 */

	FILE*	fs;
	int	status;
	int	arg;

	SetX1compat(1);

	/*yydebug = argc - 1;*/
	DisableOutput();

	initopt(argc, argv, "[old-script-file [new-script-file]] -noX1compat -constants script-files ...");
	while ((status = G_getopt(argc, argv)) == 1)
	    if (strcmp(G_optopt, "-constants") == 0)
		for (arg = 1; arg < optargc; arg++)
		  {
		    FILE*	fs;

		    CurrentFile = optargv[arg];
		    fs = fopen(CurrentFile, "r");
		    if (fs == NULL)
		      {
			perror(CurrentFile);
			return(-1);
		      }

		    InputFs(fs);
		    yyparse();
		    fclose(fs);
		  }
	    else if (strcmp(G_optopt, "-noX1compat") == 0)
		SetX1compat(0);

	if (status != 0 || optargc > 3)
	  {
	    printoptusage(argc, argv);
	    return(1);
	  }

	if (optargc > 1)
	  {
	    CurrentFile = optargv[1];
	    fs = fopen(CurrentFile, "r");
	    if (fs == NULL)
	      {
		perror(CurrentFile);
		return(1);
	      }
	  }
	else
	  {
	    CurrentFile = "<stdin>";
	    fs = stdin;
	  }

	if (optargc == 2)
	  {
	    /*
	    ** Use basename of input file as output file name
	    */

	    char*	cp;

	    for (cp = optargv[1]+strlen(optargv[1]) - 1; cp > optargv[1]; cp--)
		if (*cp == '/')
		  {
		    cp++;
		    break;
		  }

	    /*
	    ** if the input file name has no directory path then output to
	    ** stdout.
	    */

	    if (cp != optargv[1])
		freopen(cp, "w", stdout);
	  }
	else if (optargc == 3)
	    freopen(optargv[2], "w", stdout);

	EnableOutput();
	InputFs(fs);
	nextchar(1);
	yyparse();

	return 0;
}	/* convert */
